contracts/extensions/TokenSupplier.sol

```diff
@@ -0,0 +1,183 @@
+    _;
+  }
+
+  modifier isContract() {
+    uint256 size;
+    address msgSender = msg.sender;
+    assembly { size := extcodesize(msgSender) }
+    require(size > 0, "token-supplier-caller-must-be-contract");
+    _;
+  }
+
+  // Public
+
+  /// @notice Returns the identifier of the extension
+  function identifier() public override pure returns (bytes32) {
+    return keccak256("TokenSupplier");
+  }
+
+  /// @notice Returns the version of the extension
+  function version() public override pure returns (uint256) {
+    return 1;
+  }
+
+  /// @notice Configures the extension
+  /// @param _colony The colony in which the extension holds permissions
+  function install(address _colony) public override auth {
+    require(address(colony) == address(0x0), "extension-already-installed");
+
+    colony = IColony(_colony);
+  }
+
+  /// @notice Called when upgrading the extension (currently a no-op)
+  function finishUpgrade() public override auth {}
+
+  /// @notice Called when deprecating (or undeprecating) the extension (currently a no-op)
+  function deprecate(bool _deprecated) public override auth {}
+
+  /// @notice Called when uninstalling the extension
+  function uninstall() public override auth {
+    selfdestruct(address(uint160(address(colony))));
+  }
+
+  /// @notice Initialise the extension, must be called before any tokens can be issued
+  /// @param _tokenSupplyCeiling Total amount of tokens to issue
+  /// @param _tokenIssuanceRate Number of tokens to issue per day
+  function initialise(uint256 _tokenSupplyCeiling, uint256 _tokenIssuanceRate) public isRoot {
+    require(lastPinged == 0, "token-supplier-already-initialised");
+
+    tokenSupplyCeiling = _tokenSupplyCeiling;
+    tokenIssuanceRate = _tokenIssuanceRate;
+    lastPinged = block.timestamp;
+    lastRateUpdate = block.timestamp;
+  }
+
+  /// @notice Update the tokenSupplyCeiling, cannot set below current tokenSupply
+  /// @param _tokenSupplyCeiling Total amount of tokens to issue
+  function setTokenSupplyCeiling(uint256 _tokenSupplyCeiling) public isRoot isContract {
+    try ColonyExtension(msg.sender).identifier() returns (bytes32 extensionId) {
+      IColonyNetwork colonyNetwork = IColonyNetwork(colony.getColonyNetwork());
+      address installation = colonyNetwork.getExtensionInstallation(extensionId, address(colony));
+      require(installation == msg.sender, "token-supplier-not-managed-extension");
+      require(extensionId == VOTING_HYBRID, "token-supplier-cannot-set-value");
+    } catch {
+      require(false, "token-supplier-no-identifier");
+    }
+
+    tokenSupplyCeiling = _tokenSupplyCeiling;
+
+    emit TokenSupplyCeilingSet(tokenSupplyCeiling);
+  }
+
+  /// @notice Update the tokenIssuanceRate
+  /// @param _tokenIssuanceRate Number of tokens to issue per day
+  function setTokenIssuanceRate(uint256 _tokenIssuanceRate) public isRoot isContract {
```

Comment:
This function needs to issue tokens when called, otherwise the rate change is backdated.